app/src/main/java/org/p2p/wallet/moonpay/ui/new/NewBuyPresenter.kt

```diff
@@ -0,0 +1,426 @@
+import org.p2p.wallet.utils.toUsd
+import timber.log.Timber
+import java.math.BigDecimal
+
+private const val DELAY_IN_MS = 500L
+
+private val TOKENS_VALID_FOR_BUY = setOf(Constants.SOL_SYMBOL, Constants.USDC_SYMBOL)
+
+class NewBuyPresenter(
+    tokenToBuy: Token,
+    private val minBuyErrorFormat: String,
+    private val maxBuyErrorFormat: String,
+    private val buyAnalytics: BuyAnalytics,
+    private val analyticsInteractor: ScreensAnalyticsInteractor,
+    private val userInteractor: UserInteractor,
+    private val paymentMethodsInteractor: PaymentMethodsInteractor,
+    private val resourcesProvider: ResourcesProvider,
+    private val moonpayBuyInteractor: MoonpayBuyInteractor,
+    bankTransferFeatureToggle: BuyWithTransferFeatureToggle,
+) : BasePresenter<NewBuyContract.View>(), NewBuyContract.Presenter {
+
+    private lateinit var tokensToBuy: List<Token>
+
+    private val paymentMethods = mutableListOf<PaymentMethod>()
+
+    private var selectedCurrency: BuyCurrency.Currency = SelectCurrencyBottomSheet.DEFAULT_CURRENCY
+    private var selectedToken: Token = tokenToBuy
+    private lateinit var selectedPaymentMethod: PaymentMethod
+    private lateinit var currentAlphaCode: String
+
+    private var amount: String = "0"
+    private var isSwappedToToken: Boolean = false
+    private var buyDetailsState: BuyDetailsState? = null
+    private var currentBuyViewData: BuyViewData? = null
+    private val currencySelectionEnabled = bankTransferFeatureToggle.value
+
+    private var calculationJob: Job? = null
+
+    private val currenciesToSelect: List<BuyCurrency.Currency> = listOf(
+        BuyCurrency.Currency.create(Constants.GBP_SYMBOL),
+        BuyCurrency.Currency.create(Constants.EUR_SYMBOL),
+        SelectCurrencyBottomSheet.DEFAULT_CURRENCY
+    )
+
+    override fun attach(view: NewBuyContract.View) {
+        super.attach(view)
+        loadTokensToBuy()
+        loadAvailablePaymentMethods()
+    }
+
+    private fun loadTokensToBuy() {
+        launch {
+            tokensToBuy = userInteractor.getTokensForBuy(TOKENS_VALID_FOR_BUY.toList())
+            loadMoonpayBuyQuotes()
+        }
+    }
+
+    private fun loadAvailablePaymentMethods() {
+        launch {
+            // show loading
+            currentAlphaCode = paymentMethodsInteractor.getBankTransferAlphaCode()
+            val availablePaymentMethods = paymentMethodsInteractor.getAvailablePaymentMethods(currentAlphaCode)
+            selectedPaymentMethod = availablePaymentMethods.first { it.isSelected }
+            paymentMethods.addAll(availablePaymentMethods)
+            view?.showPaymentMethods(paymentMethods)
+            validatePaymentMethod()
+
+            preselectMinimalFiatAmount()
+        }
+    }
+
+    private fun loadMoonpayBuyQuotes() {
+        launch {
+            val currencyCodes = currenciesToSelect.map { it.code }
+            moonpayBuyInteractor.loadQuotes(currencyCodes, tokensToBuy)
```

Comment:
do we want to try catch some of this calls inside launch'es?

Reply:
No, it stored inside